clients/core/src/main/scala/mqperf/Receiver.scala

```diff
@@ -19,32 +20,63 @@ class Receiver(config: Config, mq: Mq, clock: Clock) extends StrictLogging {
   private val FinishWhenNoMessagesAfter = 60.seconds
 
   def run(): Future[Unit] = {
-    val mqReceiver = mq.createReceiver(config)
-    run(mqReceiver, None).flatMap(_ => mqReceiver.close())
+    val receiverFactory = mq.createReceiverFactory(config)
+    Future
+      .sequence {
+        (1 to config.receiversNumber)
+          .map(_ => {
+            receiverFactory.createReceiver()
+          }) // prepare n receivers (n = receiversNumbers)
+          .map(receiver => { // run iteration for each receiver on a separate future and close the connection afterwards
+            runIterations(receiver)
+              .flatMap(_ => receiver.close())
+          })
+      }
+      .flatMap { _ =>
+        receiverFactory.close()
+      }
+      .map(_ => ())
   }
 
-  private def run(mqReceiver: MqReceiver, lastMessage: Option[Long]): Future[Unit] = {
-    lastMessage match {
-      case Some(last) if clock.millis() - last > FinishWhenNoMessagesAfter.toMillis =>
-        logger.info(s"No messages received for ${FinishWhenNoMessagesAfter.toSeconds}s, stopping")
+  private def runIterations(mqReceiver: MqReceiver): Future[Unit] = {
+    val batchSize = config.batchSizeReceive
+
+    def receive(lastActivity: Long): Future[Unit] = {
+      if (clock.millis() - lastActivity > FinishWhenNoMessagesAfter.toMillis) {
         Future.successful(())
-      case _ =>
+      } else {
         mqReceiver
-          .receive(config.batchSizeReceive)
-          .flatMap { msgs =>
+          .receive(batchSize)
+          .map { msgs =>
             val now = clock.millis()
             msgs.foreach { case (_, msg) =>
               val t = Timestamp.extract(msg)
               LocalMetrics.messageLatencyHistogram.observe((now - t).toDouble)
             }
             LocalMetrics.messageCounter.inc(msgs.size.toDouble)
-
-            mqReceiver.ack(msgs.map(_._1)).map(_ => msgs.size)
+            mqReceiver
```

Comment:
maybe we should add a comment that this intentionally runs asynchronously

Reply:
good idea, done